packages/labs/router/src/routes.ts

```diff
@@ -0,0 +1,212 @@
+        childRoutes.goto(tailGroup);
+      }
+    }
+    this._host.requestUpdate();
+  }
+
+  /**
+   * The result of calling the current route's render() callback.
+   */
+  get outlet() {
+    return this._currentRoute?.render(this._currentParams);
+  }
+
+  /**
+   * The current parsed route parameters.
+   */
+  get params() {
+    return this._currentParams;
+  }
+
+  /**
+   * Matches `url` against the installed routes and returns the first match.
+   */
+  private _getRoute(pathname: string) {
+    return this._routes.find((r) => r.pattern.test({pathname: pathname}));
+  }
+
+  hostConnected() {
+    this._host.addEventListener(
+      RoutesConnectedEvent.eventName,
+      this._onRoutesConnected
+    );
+    // TODO: connect to parent router
+    const event = new RoutesConnectedEvent(this);
+    this._host.dispatchEvent(event);
+    this._onDisconnect = event.onDisconnect;
+  }
+
+  hostDisconnected() {
+    // When this child routes constroller is disconnected because a parent
+    // outlet rendered a different template, disconnecting will ensure that
+    // this controller doesn't receive a tail match meant for another route.
+    this._onDisconnect?.();
+  }
+
+  private _onRoutesConnected = (e: RoutesConnectedEvent) => {
+    // Don't handle the event fired by this routes controller, which we get
+    // because we do this.dispatchEvent(...)
+    if (e.routes === this) {
+      return;
+    }
+
+    const childRoutes = e.routes;
+    this._childRoutes.push(childRoutes);
+
+    e.stopImmediatePropagation();
+    e.onDisconnect = () => {
+      // Remove route from this._childRoutes:
+      this._childRoutes?.splice(
+        this._childRoutes.indexOf(childRoutes) >>> 0,
+        1
+      );
+    };
+
+    const tailGroup = getTailGroup(this._currentParams);
+    if (tailGroup !== undefined) {
+      childRoutes.goto(tailGroup);
+    }
+  };
+}
+
+/**
+ * Returns the tail of a pathname groups object. This is the match from a
+ * wildcard at the end of a pathname pattern, like `/foo/*`
+ */
+const getTailGroup = (groups: {[key: string]: string}) => {
+  let tailKey: string | undefined;
+  for (const key of Object.keys(groups)) {
+    if ((/\d+/.test(key) && tailKey === undefined) || key > tailKey!) {
+      tailKey = key;
+    }
+  }
+  return tailKey && groups[tailKey];
+};
+
+/**
+ * This event is fired from Routes controllers when their host is connected to
+ * announce the child route and potentially connect to a parent routes controller.
+ */
+export class RoutesConnectedEvent extends Event {
+  static readonly eventName = 'routes-connected';
+  readonly routes: Routes;
+  onDisconnect?: () => void;
+
+  constructor(routes: Routes) {
+    super(RoutesConnectedEvent.eventName, {
+      bubbles: true,
+      composed: true,
+      cancelable: false,
+    });
+    this.routes = routes;
+  }
+}
+
+declare global {
+  interface HTMLElementEventMap {
+    'routes-connected': RoutesConnectedEvent;
```

Comment:
No strong opinion, but should we use `lit-routes-connected` for a lower chance of event name collision? We use `lit-localize-status` in localize.

Reply:
ah, good point